// @nuxthq/studio/dist/theme.cjs
'use strict';

const defu = require('defu');

const supportedFields = {
  /**
   * Raw types
   */
  default: {
    type: "string",
    tags: [
      "@studioInput string"
    ]
  },
  string: {
    type: "string",
    tags: [
      "@studioInput string"
    ]
  },
  number: {
    type: "number",
    tags: [
      "@studioInput number"
    ]
  },
  boolean: {
    type: "boolean",
    tags: [
      "@studioInput boolean"
    ]
  },
  array: {
    type: "array",
    tags: [
      "@studioInput array"
    ]
  },
  "design-token": {
    type: "string",
    tags: [
      "@studioInput design-token"
    ]
  },
  object: {
    type: "object",
    tags: [
      "@studioInput object",
      "@studioInputObjectValueType icon"
    ]
  },
  file: {
    type: "string",
    tags: [
      "@studioInput file"
    ]
  },
  component: {
    type: "string",
    tags: [
      "@studioInput component"
    ]
  },
  icon: {
    type: "string",
    tags: [
      "@studioInput icon"
    ]
  },
  /**
   * Design Tokens
   */
  color: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType color"
    ]
  },
  size: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType size"
    ]
  },
  shadow: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType shadow"
    ]
  },
  opacity: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType opacity"
    ]
  },
  font: {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType font"
    ]
  },
  "font-weight": {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType font-weight"
    ]
  },
  "font-size": {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType size"
    ]
  },
  "letter-spacing": {
    type: "string",
    tags: [
      "@studioInput design-token",
      "@studioInputTokenType size"
    ]
  }
};
function field(type, defaultValue) {
  const result = defu.defu(
    supportedFields[typeof type === "string" ? type : type.type],
    type
  );
  if (!result.tags) {
    result.tags = [];
  }
  if (result.icon) {
    result.tags.push(`@studioIcon ${result.icon}`);
    delete result.icon;
  }
  if (defaultValue) {
    result.default = defaultValue;
  }
  return result;
}

exports.field = field;
