// pinceau/dist/utils.mjs
import { c as castVariants, d as castVariantsPropsAst, e as evalCssDeclaration, f as pushVariantsProps, r as resolveCssCallees, g as resolveRuntimeContents, h as resolveVariantsProps, t as transformCssFunction, i as transformVariants } from './shared/pinceau.23c34d4f.mjs';
export { J as JS_EXTENSIONS, S as STYLES_EXTENSIONS, A as SUPPORTED_EXTENSIONS, q as astTypes, x as debugMarker, n as defaultExport, z as errorMessage, l as expressionToAst, y as fileLink, s as findLineColumn, w as getDebugContext, m as merger, b as message, j as parseAst, p as parsePinceauQuery, k as parseVueComponent, o as printAst, a as updateDebugContext, u as useDebugPerformance, v as visitAst } from './shared/pinceau.23c34d4f.mjs';
import { b as replaceStyleTs, e as resolveScriptSetup, f as resolveStyle, c as resolveStyleQuery, g as resolveTemplate, h as transformAddPinceauClass, i as transformAddPropsKey, j as transformAddRuntimeImports, k as transformComputedStyles, t as transformDtHelper, m as transformFinishRuntimeSetup, n as transformMediaQueries, o as transformScheme, q as transformStyle, d as transformVueSFC, R as RESOLVED_ID_RE, x as cssFull, y as definitionsFull, z as generateTheme, A as importConfigFile, B as loadLayers, p as prepareOutputDir, C as resolveConfigFile, D as resolveConfigLayer, E as resolveConfigSources, F as resolveDefinitions, G as resolveMediaQueriesKeys, H as schemaFull, I as stubOutputs, J as tsFull, K as usePinceauConfigContext, u as usePinceauContext, L as usePinceauVirtualStore, M as utilsFull } from './shared/pinceau.aa6b702d.mjs';
export { w as flattenTokens, s as isSafeConstName, l as loadVueStyle, v as objectPaths, r as registerAliases, a as registerPostCssPlugins } from './shared/pinceau.aa6b702d.mjs';
import { d as defineTheme } from './shared/pinceau.5e8d4555.mjs';
export { D as DARK, I as INITIAL, L as LIGHT, d as castValue, b as castValues, c as createTokensHelper, l as darkRegex, u as dtRegex, h as get, y as getResolvedSelectors, k as keyRegex, q as lightRegex, j as mqCssRegex, m as mqPlainRegex, n as normalizeConfig, o as outputFileNames, p as pathToVarName, i as referencesRegex, a as resolveCssProperty, f as resolveCustomDirectives, e as resolveReferences, g as resolveThemeRule, r as resolveVariableFromPath, x as responsiveMediaQueryRegex, s as set, z as stringify, t as transformTokensToVariable, v as variantsRegex, w as walkTokens } from './shared/pinceau.3988746b.mjs';
import 'defu';
import 'acorn';
import 'perf_hooks';
import 'ohash';
import 'scule';
import 'fs';
import 'fs/promises';
import 'jiti';
import 'pathe';
import 'vue';
import 'style-dictionary-esm';
import 'untyped';
import 'postcss-nested';
import 'postcss-custom-properties';
import 'postcss-dark-theme-class';
import 'vue/compiler-sfc';
import 'recast';
import 'recast/parsers/typescript.js';

const index$1 = {
	__proto__: null,
	castVariants: castVariants,
	castVariantsPropsAst: castVariantsPropsAst,
	evalCssDeclaration: evalCssDeclaration,
	pushVariantsProps: pushVariantsProps,
	replaceStyleTs: replaceStyleTs,
	resolveCssCallees: resolveCssCallees,
	resolveRuntimeContents: resolveRuntimeContents,
	resolveScriptSetup: resolveScriptSetup,
	resolveStyle: resolveStyle,
	resolveStyleQuery: resolveStyleQuery,
	resolveTemplate: resolveTemplate,
	resolveVariantsProps: resolveVariantsProps,
	transformAddPinceauClass: transformAddPinceauClass,
	transformAddPropsKey: transformAddPropsKey,
	transformAddRuntimeImports: transformAddRuntimeImports,
	transformComputedStyles: transformComputedStyles,
	transformCssFunction: transformCssFunction,
	transformDtHelper: transformDtHelper,
	transformFinishRuntimeSetup: transformFinishRuntimeSetup,
	transformMediaQueries: transformMediaQueries,
	transformScheme: transformScheme,
	transformStyle: transformStyle,
	transformVariants: transformVariants,
	transformVueSFC: transformVueSFC
};

const index = {
	__proto__: null,
	RESOLVED_ID_RE: RESOLVED_ID_RE,
	cssFull: cssFull,
	defineTheme: defineTheme,
	definitionsFull: definitionsFull,
	generateTheme: generateTheme,
	importConfigFile: importConfigFile,
	loadLayers: loadLayers,
	prepareOutputDir: prepareOutputDir,
	resolveConfigFile: resolveConfigFile,
	resolveConfigLayer: resolveConfigLayer,
	resolveConfigSources: resolveConfigSources,
	resolveDefinitions: resolveDefinitions,
	resolveMediaQueriesKeys: resolveMediaQueriesKeys,
	schemaFull: schemaFull,
	stubOutputs: stubOutputs,
	tsFull: tsFull,
	usePinceauConfigContext: usePinceauConfigContext,
	usePinceauContext: usePinceauContext,
	usePinceauVirtualStore: usePinceauVirtualStore,
	utilsFull: utilsFull
};

export { index as theme, index$1 as transforms };
