// @nuxtjs/plausible/dist/module.mjs
import { fileURLToPath } from 'node:url';
import { join } from 'pathe';
import { defu } from 'defu';
import { defineNuxtModule, addImportsDir, addPlugin } from '@nuxt/kit';

const name = "@nuxtjs/plausible";
const version = "0.2.1";

const module = defineNuxtModule({
  meta: {
    name,
    version,
    configKey: "plausible",
    compatibility: {
      nuxt: "^3"
    }
  },
  defaults: {
    hashMode: false,
    trackLocalhost: false,
    domain: "",
    apiHost: "https://plausible.io",
    autoPageviews: true,
    autoOutboundTracking: false
  },
  setup(options, nuxt) {
    nuxt.options.runtimeConfig.public.plausible = defu(
      nuxt.options.runtimeConfig.public.plausible,
      options
    );
    const runtimeDir = fileURLToPath(new URL("./runtime", import.meta.url));
    nuxt.options.build.transpile.push(runtimeDir);
    addImportsDir(join(runtimeDir, "composables"));
    addPlugin({
      src: join(runtimeDir, "plugin.client"),
      mode: "client"
    });
  }
});

export { module as default };
